release-lib-master.Jenkinsfile

```diff
@@ -0,0 +1,189 @@
+    PIPELINE_USERNAME = sh(returnStdout: true, script: 'id -un').trim()
+  }
+  stages {
+    stage('git checkout') {
+      steps{
+        script{
+          checkout([$class      : 'GitSCM', branches: [[name: params.LIB_GIT_BRANCH]], doGenerateSubmoduleConfigurations: false,
+                    extensions  : [[$class: 'RelativeTargetDirectory', relativeTargetDir: params.LIB_NAME]], submoduleCfg: [],
+                    userRemoteConfigs: [[url: params.LIB_GIT_REPO]]])
+
+          echo "Prepare the release of ${params.LIB_GIT_REPO} branch: ${params.LIB_GIT_BRANCH}"
+
+          sh """#!/bin/bash
+          cd ${params.LIB_NAME}
+          git checkout ${params.LIB_GIT_BRANCH}
+          gpg --allow-secret-key-import --import /home/jenkins/gnupg_keys/private_key.txt
+          """
+        }
+      }
+    }
+    stage('Setup and Formating'){
+      steps{
+        script{
+          withCredentials([
+            usernamePassword(credentialsId:'GitHub-Bot', passwordVariable:'BOT_PASSWORD', usernameVariable:'BOT_USERNAME'),
+            usernamePassword(credentialsId:'OSS-Nexus-Bot', passwordVariable:'OSS_BOT_PASSWORD', usernameVariable:'OSS_BOT_USERNAME'),
+            string(credentialsId: 'gnupg_passphrase', variable: 'PASSPHRASE')
+          ]) {
+            dir(params.LIB_NAME){
+              sh """
+              git config --global user.email "${params.BOT_EMAIL}"
+              git config --global user.name "${BOT_USERNAME}"
+              mvn -B -s ../settings.xml -Pformatting clean install
+              git commit -a, "Update license header"                #commit nothing when there is no file needs to be modified
+              git push https://${BOT_USERNAME}:${BOT_PASSWORD}@$`python3 -c 'print("${params.LIB_GIT_REPO}".split("//")[1])'` --all
+              cp ../settings-release.xml /home/jenkins/.m2/settings.xml
+              sed -i 's/{{_PASSPHRASE}}/${PASSPHRASE}/g' /home/jenkins/.m2/settings.xml
+              sed -i 's/{{_USERNAME}}/${OSS_BOT_USERNAME}/g' /home/jenkins/.m2/settings.xml
+              sed -i 's/{{_PASSWORD}}/${OSS_BOT_PASSWORD}/g' /home/jenkins/.m2/settings.xml
+              """
+            }
+          }
+        }
+      }
+    }
+    stage('Release Prepare'){
+      steps{
+        script{
+          dir(params.LIB_NAME){
+            env.LIB_NEXT_VERSION = sh (
+              script: """
+                echo ${params.LIB_MAJOR_VERSION} | awk -F. -v OFS=. 'NF==1{print ++$NF}; NF>1{if(length($NF+1)>length($NF))$(NF-1)++; $NF=sprintf("%0*d", length($NF), ($NF+1)%(10^length($NF))); print}'
+                """
+              returnStdout: true
+            )
+            sh """#!/bin/bash
+            mvn help:effective-settings
+            mvn --batch-mode release:prepare -DreleaseVersion=${params.LIB_MAJOR_VERSION} -DdevelopmentVersion=${env.LIB_NEXT_VERSION}-SNAPSHOT -Dtag=indy-parent-${params.LIB_MAJOR_VERSION}
```

Comment:
I think it's ok to use the default tag name. That will make this pipeline script more general-purpose.

Reply:
Also, do you need a SSH key registered for commonjava-bot so you can push the tag back up to developerConnection? I'm not sure whether pushing over https will work, I've never tried it.

Reply:
the tag name should be the same as default one, but I have to claim it so it won't ask for input.
you can check the latest version of pipeline, which uses bot's username and password in github as authentication, of course it is jenkins credential to prevent leak-out .